### PR TITLE
File upload support from the editor

### PR DESCRIPTION
**Pull request by vishaltelangre**
Fixes https://github.com/bigbinary/acehelp/issues/626.

### CODE REVIEW

File: app/javascript/packs/admin.js [hunk 1]
```diff
@@ -67,13 +67,50 @@ document.addEventListener("DOMContentLoaded", () => {
...
-    document.addEventListener("trix-attachment-add", function(event) {
-        var attachment;
-        attachment = event.attachment;
-        if (attachment.file) {
-            //return uploadAttachment(attachment);
-            //console.log(attachment.file);
+    function uploadAttachment(apiKey, attachment, attachmentsPath) {
+        const file = attachment.file;
+        const form = new FormData();
+        const xhr = new XMLHttpRequest();
```
Comment (vishaltelangre) [inline app/javascript/packs/admin.js]: JavaScript's Fetch API does not seem to allow capturing the upload progress [1], therefore we will stick to making a traditional XHR request instead. 

--
[1] https://stackoverflow.com/questions/36453950/upload-file-with-fetch-api-in-javascript-and-show-progress

**Comment by vishaltelangre** [comment]
### Update: 

File upload functionality works from the editor on the edit article page.

This functionality will also work from the editor on the new article page once the related pull request https://github.com/bigbinary/acehelp/pull/648 becomes ready, which @tejaswinichile is working on.

cc: @neerajdotname @ritesh404

File: app/javascript/Admin/Page/Article/Edit.elm [hunk 2]
```diff
@@ -412,10 +422,17 @@ view model =
                     ]
                 , div
                     [ class "row article-content" ]
-                    [ div [ classList [ ( "hidden", not model.isEditable ) ] ]
-                        [ node "trix-editor"
+                    [ div
+                        [ classList [ ( "hidden", not model.isEditable ) ]
+                        , attribute "data-attachments-path" model.attachmentsPath
+                        , attribute "data-api-key" orgKey
```
Comment (vishaltelangre) [inline app/javascript/Admin/Page/Article/Edit.elm]: We use these custom data-attributes while issuing an XHR request to upload attachments.

File: app/javascript/packs/admin.js [hunk 3]
```diff
@@ -62,18 +63,48 @@ document.addEventListener("DOMContentLoaded", () => {
...
+    // for the each file separately.
+    app.ports.addAttachments.subscribe(() => {
+        const editorNode = document.querySelector("trix-editor");
+        const fileInput = document.createElement("input");
+
+        fileInput.setAttribute("type", "file");
+        fileInput.setAttribute("multiple", "");
+
+        fileInput.addEventListener("change", function(event) {
+            const files = Array.from(this.files);
```
Comment (vishaltelangre) [inline app/javascript/packs/admin.js]: We cannot use the map function on the `this.files` object directly, so we convert it to an Array first.

**Comment by vishaltelangre** [comment]
@ritesh404 _a This is ready now. Please review.

File: app/javascript/packs/attachment_uploader.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            : this.uploadErrorHandler();
+    }
+
+    uploadSuccessHandler() {
+        const { responseText } = this.request;
+        const { attachment_url } = JSON.parse(responseText);
+
+        return this.attachment.setAttributes({
+            url: attachment_url,
+            href: attachment_url
```
Comment (ritesh404) [inline app/javascript/packs/attachment_uploader.js]: Is this all that we need to do to update UI?

Comment (vishaltelangre) [reply app/javascript/packs/attachment_uploader.js]: Yes. Trix editor adds a markup similar to the below snippet to the contents using these details.

```html
<a href="attachment_href_here">
  <img src="attachment_url_here" />
</a>
```

**Comment by ritesh404** [comment]
On deleting an article we are deleting the associated images with it.
What happens when we edit the article and just remove the image?

**Comment by ritesh404** [comment]
@vishaltelangre _a can you fix the issues mentioned above

**Comment by vishaltelangre** [comment]
> What happens when we edit the article and just remove the image?

This PR only addresses the upload part. We will be addressing the actual deletion functionality of the file when clicking on the delete icon on it separately.

**Comment by vishaltelangre** [comment]
@ritesh404 _a Please review again. I have fixed both of the issues which you had reported.

**Comment by ritesh404** [comment]
@vishaltelangre _a please fix the merge conflict and merge

**Comment by vishaltelangre** [comment]
Rebased and fixed merge conflicts.

**Pull request merged**